doc/examples/applications/plot_3d_segmentation_embryo.py

```diff
@@ -0,0 +1,182 @@
+ax[0].imshow(smooth[25, :, :])
+ax[0].set_title('Smoothing out')
+ax[0].axis('off')
+ax[1].imshow(regions[25, :, :])
+ax[1].set_title('Multi-Otsu thresholding')
+ax[1].axis('off')
+
+#####################################################################
+# We identify nuclei to be the brightest of the three classes and we remove
+# small objects.
+
+cells_noisy = smooth > thresholds[1]
+cells = ski.morphology.opening(cells_noisy, footprint=np.ones((3, 5, 5)))
+
+#####################################################################
+# We use the watershed algorithm to separate nuclei when they are touching
+# or overlapping.
+
+distance = ndi.distance_transform_edt(cells)
+
+local_max_coords = ski.feature.peak_local_max(distance, min_distance=12)
+local_max_mask = np.zeros(distance.shape, dtype=bool)
+local_max_mask[tuple(local_max_coords.T)] = True
+markers = ski.measure.label(local_max_mask)
+
+segmented_cells = ski.segmentation.watershed(-distance, markers, mask=cells)
+
+fig, ax = plt.subplots(ncols=2, figsize=(10, 5))
+ax[0].imshow(cells[25, :, :], cmap='gray')
+ax[0].set_title('Touching nuclei')
+ax[0].axis('off')
+ax[1].imshow(ski.color.label2rgb(segmented_cells[25, :, :], bg_label=0))
+ax[1].set_title('Segmented nuclei')
+ax[1].axis('off')
+
+#####################################################################
+# With the naked eye, we can see a slight over-segmentation... How does this
+# segmentation compare with that obtained in [1]_?
+
+#####################################################################
+# We used the software developed for the research project at:
+# (https://bitbucket.org/fernandoamat/tgmm-paper/src/master/doc/new/docs/user-guide/quickstart.md)
+# We edited the TGMM config file to apply the hierarchical segmentation on
+# the sample data, which we saved 'back' in KLB format:
+#
+# .. code-block:: python
+#
+#     klb.writefull(np.ascontiguousarray(sample), 'sample_3D_frame_184.klb')
+#
+# We installed the software (https://bitbucket.org/fernandoamat/tgmm-paper/src/master/doc/new/docs/dev-guide/building.md)
+# and ran it:
+#
+# .. code-block:: bash
+#
+#     ProcessStack config.md 184
+#     ProcessStack sample_3D_frame_184_seg_conn74_rad2.bin 14 50
+#
+# Chose tau=14 because persistanceSegmentationTau=14 in config file.
+# A value of tau=2 clearly yields over-segmentation (yields 1597 nuclei).
+# Chose minSuperVoxelSzPx=50 because minNucleiSize=50 in config file.
+# Not much difference between minSuperVoxelSzPx=50 (yiels 517 nuclei) and
+# minSuperVoxelSzPx=14 (yiels 541 nuclei).
+# The output (segmentation result) is a KLB file; we save it into a Numpy array.
+
+res_url = 'https://github.com/mkcor/draft-notebooks/blob/main/embryo/data/tgmm_conn74_tau14.npy'
+resp = requests.get(res_url)
+gt = np.load(io.BytesIO(resp.content))
+
+# Ensure TGMM result is an image of type "labeled"
+assert gt.dtype in [np.uint16, np.uint32, np.uint64]
+assert gt.min() == 0
+assert gt.max() == np.unique(gt).shape[0] - 1
```

Comment:
@lagru maybe, once we type the API, the above could turn into a one-liner looking like `assert isinstance(gt, Labels)`? :muscle:

Reply:
I'm not sure, `isinstance()` works with class inheritance, typing is not necessarily about that but might be about structural typing and such. And I find it difficult to imagine that running a type checker would check the actual content like it's done here.

Reply:
What is *structural typing*? :eyes:

Reply:
Have a look at [protocols](https://typing.readthedocs.io/en/latest/spec/protocol.html) and [PEP 544 -- Protocols: Structural subtyping (static duck typing)](https://peps.python.org/pep-0544/) if you are interested. Basically it's trying to address the problem that typing doesn't support a lot of the more dynamic duck typing capabilities of Python. It's also referred to as "static duck typing".